storage/storage.go

```diff
@@ -47,38 +46,82 @@ func NewClient(url *objurl.ObjectURL) (Storage, error) {
 
 // Object is a generic type which contains metadata for storage items.
 type Object struct {
-	URL          *objurl.ObjectURL
-	Etag         string
-	ModTime      time.Time
-	Mode         os.FileMode
-	Size         int64
-	StorageClass StorageClass
-	Err          error
+	URL          *objurl.ObjectURL `json:"key,omitempty"`
+	Etag         string            `json:"etag,omitempty"`
+	ModTime      *time.Time        `json:"last_modified,omitempty"`
+	Type         ObjectType        `json:"type,omitempty"`
+	Size         int64             `json:"size,omitempty"`
+	StorageClass StorageClass      `json:"storage_class,omitempty"`
+	Err          error             `json:"error,omitempty"`
 }
 
 // String returns the string representation of Object.
 func (o *Object) String() string {
 	return o.URL.String()
 }
 
+// JSON returns the JSON representation of Object.
+func (o *Object) JSON() string {
+	bytes, _ := json.Marshal(o)
+	return string(bytes)
+}
+
+// ObjectType is the type of Object.
+type ObjectType struct {
+	mode os.FileMode
+}
+
+// String returns the string representation of ObjectType.
+func (o ObjectType) String() string {
+	switch mode := o.mode; {
+	case mode.IsRegular():
+		return "file"
+	case mode.IsDir():
+		return "directory"
+	case mode&os.ModeSymlink != 0:
+		return "symlink"
+	}
+	return ""
+}
+
+// MarshallJSON returns the stringer of ObjectType as a marshalled json.
+func (o ObjectType) MarshalJSON() ([]byte, error) {
+	return json.Marshal(o.String())
+}
+
+// IsDir checks if the object is a directory.
+func (o ObjectType) IsDir() bool {
+	return o.mode.IsDir()
+}
+
+// dateFormat is a constant time template for the bucket.
+const dateFormat = "2006/01/02 15:04:05"
+
 // Bucket is a container for storage objects.
 type Bucket struct {
-	CreationDate time.Time
-	Name         string
+	CreationDate time.Time `json:"created_at"`
+	Name         string    `json:"name"`
 }
 
 // String returns the string representation of Bucket.
 func (b Bucket) String() string {
 	return fmt.Sprintf("%s  s3://%s", b.CreationDate.Format(dateFormat), b.Name)
 }
 
+// String returns the JSON representation of Bucket.
+func (b Bucket) JSON() string {
+	bytes, _ := json.Marshal(b)
```

Comment:
this pattern is used a lot of times. we can write function like `jsonMarshal(v interface{}) string`

Reply:
Where should it be located? Both `storage`, `stats` and `core` uses this. I could not find any good place to do this.

Reply:
I'm not sure. All packages may hold their own jsonMarshal() function.

Reply:
Done. Let's use `jsonMarshal` helper function on `core` and `storage` packages. I did not update `stats` since it uses this pattern only once.